src/main/java/nextstep/subway/domain/LineSections.java

```diff
@@ -126,41 +123,36 @@ public List<Station> getOrderedStations() {
     }
 
     private Section findBothEndsStations() {
-        List<Station> upStations = this.sections.stream()
+        Set<Station> upStations = this.sections.stream()
             .map(Section::getUpStation)
-            .collect(Collectors.toList());
-        List<Station> downStations = this.sections.stream()
+            .collect(Collectors.toSet());
+        Set<Station> downStations = this.sections.stream()
             .map(Section::getDownStation)
-            .collect(Collectors.toList());
+            .collect(Collectors.toSet());
 
-        Station upEndStation = upStations.stream()
-            .filter(s -> !downStations.contains(s))
-            .findFirst()
-            .orElse(null);
-        Station downEndStation = downStations.stream()
-            .filter(s -> !upStations.contains(s))
-            .findFirst()
-            .orElse(null);
+        HashSet<Station> allMiddleStations = getAllMiddleStations(upStations, downStations);
+
+        upStations.removeAll(allMiddleStations);
+        downStations.removeAll(allMiddleStations);
 
         return Section.builder()
-            .upStation(upEndStation)
-            .downStation(downEndStation)
+            .upStation(upStations.stream().findFirst().orElse(null))
+            .downStation(downStations.stream().findFirst().orElse(null))
```

Comment:
```suggestion
            .upStation(getFirstStation())
            .downStation(getLastStation())
```

이런식으로 메서드 분리를 해봐도 좋을 것 같네요 👍